go/chat/search/indexer.go

```diff
@@ -548,115 +607,183 @@ func (idx *Indexer) Search(ctx context.Context, uid gregor1.UID, query string, o
+				break
+			}
+			if opts.MaxConvsHit > 0 && len(hitMap) >= opts.MaxConvsHit {
+				idx.Debug(ectx, "Search: max hit convs reached, ending search")
+				break
 			}
 		}
-	}
+		return nil
+	})
 
-	var numConvsSearched, numConvsHit int
-	hits := []chat1.ChatSearchInboxHit{}
-	convList := idx.flattenConvMap(ctx, uid, convMap)
-	idx.Debug(ctx, "Search: beginning conv searches")
-	for _, conv := range convList {
-		select {
-		case <-ctx.Done():
-			return nil, ctx.Err()
-		default:
-		}
-		convID := conv.GetConvID()
-		convIDStr := convID.String()
-		numConvsSearched++
-		convIdx := convIdxMap[convIDStr]
-		msgIDs, err := idx.searchConv(ctx, convID, convIdx, uid, tokens, opts)
-		if err != nil {
-			return nil, err
-		}
-		convHits, err := idx.searchHitsFromMsgIDs(ctx, conv, uid, msgIDs, queryRe, opts)
-		if err != nil {
-			return nil, err
-		}
-		if len(msgIDs) != convHits.Size() {
-			idx.Debug(ctx, "Search: hit mismatch, found %d msgIDs in index, %d hits in conv: %v",
-				len(msgIDs), convHits.Size(), conv.GetName())
-		}
-		if convHits == nil {
-			continue
-		}
-		convHits.Query = query
-		numConvsHit++
-		if hitUICh != nil {
-			// Stream search hits back to the UI channel
-			hitUICh <- *convHits
-		}
-		hits = append(hits, *convHits)
-		if opts.MaxConvsSearched > 0 && numConvsSearched >= opts.MaxConvsSearched {
-			idx.Debug(ctx, "Search: max search convs reached, ending search")
-			break
-		}
-		if opts.MaxConvsHit > 0 && numConvsHit >= opts.MaxConvsHit {
-			idx.Debug(ctx, "Search: max hit convs reached, ending search")
-			break
-		}
+	if err := eg.Wait(); err != nil {
+		return nil, err
 	}
-	// kick this off in the background after we have our results so there is no
-	// lock contention during the search
+
 	switch opts.ReindexMode {
-	case chat1.ReIndexingMode_AFTERSEARCH:
+	case chat1.ReIndexingMode_POSTSEARCH_ASYNC,
+		chat1.ReIndexingMode_POSTSEARCH_SYNC:
 		for _, conv := range convList {
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			default:
+			}
 			convIDStr := conv.GetConvID().String()
 			convIdx := convIdxMap[convIDStr]
-			_, _, err = idx.reindexConv(ctx, conv.Conv, uid, convIdx, reindexOpts{forceReindex: false})
-			if err != nil {
-				return nil, err
+
+			switch opts.ReindexMode {
+			// kick this off in the background after we have our results so there is no
+			// lock contention during the search
+			case chat1.ReIndexingMode_POSTSEARCH_ASYNC:
+				_, _, err = idx.reindexConv(ctx, conv.Conv, uid, convIdx, reindexOpts{forceReindex: false})
+				if err != nil {
+					return nil, err
+				}
+			// reindex all of the conversations and research to populate any new results
+			case chat1.ReIndexingMode_POSTSEARCH_SYNC:
+				// ignore any fully indexed convs since we respect
+				// opts.MaxConvsSearched
+				if convIdx.PercentIndexed(conv.Conv) == 100 {
+					continue
+				}
+				convIdx, totalPercentIndexed, err = idx.reindexConvWithUIUpdate(ctx, conv.Conv, uid,
```

Comment:
I wonder if this whole thing could be in another goroutine in the errgroup.

Reply:
yeah that sounds like a good idea, i'll leave the postsearch async as is and rework post search sync to be in a thread

Reply:
punting on this for simplicity, main benefit is when initially read out the indices from disk